src/index.ts

```diff
@@ -1685,130 +1867,163 @@ type DurationValue = {
+	$time?: FilterFunctionValue<T>;
+	$totaloffsetminutes?: FilterFunctionValue<T>;
+	$now?: FilterFunctionValue<T>;
 	$duration?: DurationValue;
-	$maxdatetime?: FilterFunctionValue;
-	$mindatetime?: FilterFunctionValue;
-	$totalseconds?: FilterFunctionValue;
-	$round?: FilterFunctionValue;
-	$floor?: FilterFunctionValue;
-	$ceiling?: FilterFunctionValue;
-	$isof?: FilterFunctionValue;
-	$cast?: FilterFunctionValue;
-}
+	$maxdatetime?: FilterFunctionValue<T>;
+	$mindatetime?: FilterFunctionValue<T>;
+	$totalseconds?: FilterFunctionValue<T>;
+	$round?: FilterFunctionValue<T>;
+	$floor?: FilterFunctionValue<T>;
+	$ceiling?: FilterFunctionValue<T>;
+	$isof?: FilterFunctionValue<T>;
+	$cast?: FilterFunctionValue<T>;
+};
+export type FilterObj<T extends Resource['Read'] = AnyObject> = {
+	[key in StringKeyOf<T>]?: Filter<T> | Lambda<T> | undefined;
+} & FilterOperations<T>;
 
-export type FilterArray = Filter[];
+export type FilterArray<T extends Resource['Read'] = AnyObject> = Array<
+	Filter<T>
+>;
 export type FilterBaseType = string | number | null | boolean | Date;
-export type RawFilter =
+export type RawFilter<T extends Resource['Read'] = AnyObject> =
 	| string
-	| [string, ...Filter[]]
+	| [string, ...Array<Filter<T>>]
 	| {
 			$string: string;
-			[index: string]: Filter;
+			[index: `${Letter | Digit}${string}`]: Filter<T>;
 	  };
-export interface Lambda {
-	$alias: string;
-	$expr: Filter;
-}
-export type Filter = FilterObj | FilterArray | FilterBaseType;
 
-export type ResourceExpand = Dictionary<ODataOptions>;
+// TODO: Lambdas need a real re-think because we need to know the property name, eg `belongs_to__application: { $any: {... } }` needs T to become `Application` rather than `Device`
+export interface Lambda<
+	T extends Resource['Read'] = AnyObject,
+	Alias extends string = string,
+> {
+	$alias: Alias;
+	$expr: Filter<T & { [a in Alias]: T }>;
+}
+export type Filter<T extends Resource['Read'] = AnyObject> =
+	| FilterObj<T>
+	| FilterArray<T>
+	| FilterBaseType;
+
+export type ResourceExpand<T extends Resource['Read'] = AnyObject> = {
+	[resource in StringKeyOf<T> as ExtractExpand<T, resource> extends never
+		? never
+		: resource]?: ODataOptions<ExtractExpand<T, resource>>;
+};
 
-export type Expand = string | ResourceExpand | Array<string | ResourceExpand>;
+export type BaseExpand<T extends Resource['Read'] = AnyObject> =
+	| ExpandableStringKeyOf<T>
+	| ResourceExpand<T>;
+export type Expand<T extends Resource['Read'] = AnyObject> =
+	| BaseExpand<T>
+	| Array<BaseExpand<T>>;
 
 type OrderByDirection = 'asc' | 'desc';
 
-export type OrderBy =
-	| string
-	| OrderBy[]
-	| {
-			[k: StartsWithLetter]: OrderByDirection;
-	  }
+export type OrderBy<T extends Resource['Read'] = AnyObject> =
+	| StringKeyOf<T>
+	| Array<OrderBy<T>>
+	| { [k in StringKeyOf<T>]?: OrderByDirection }
 	| ({
-			[k: StartsWithLetter]: {
-				$count: ODataCountOptions;
+			[k: `${Letter}${string}`]: {
+				$count: ODataCountOptions<T>;
```

Comment:
Shouldn't this be v ?
```suggestion
				$count: ODataCountOptions<T[k]>;
```
Which means we might need to do
```
[k in (StringKeyOf<T> & `${Letter}${string}`)]
```
or something.

Reply:
Had to jump through a bunch of hoops to get this working properly but now it is 🎉